### ISSUE TITLE
Break up the CMP into separate documents

### ISSUE THREAD

**Issue by erco77** [erco77/fltk#2]
So I'm working on separating the CMP into at least a separate document for the build related docs.

I'm starting with moving it all to a README.developer.txt file, and if that becomes too restrictive, we can always move it to some higher level document format, like Doxygen or some such.

I decided to start with the build macros; attaching [README.developer.txt](https://github.com/erco77/fltk/files/10695265/README.developer.txt)

If this seems OK, I'll move forward and modify the CMP section on macro variables to simply refer to this.

**Comment by Albrecht-S** [comment]
@erco77 IMHO `README.developer.txt` is a good start. Its final destination could be the developer documentation that's currently being created: Manolo's Wayland docs and more (in Doxygen format). I'm going to split this in two distinct documents as discussed ("User Docs" and "Developer Docs") but that's not yet done. Finally the contents of `README.developer.txt` could (should ?) be included in these docs.

I looked through your draft and made some edits, see attached [fixes.diff](https://github.com/erco77/fltk/files/10697975/fixes.diff.txt). My modifications are:
1. remove `__APPLE_QUARTZ__` which is no longer used
2. adjust overview and paragraph headers according to (1)
3. fix `FL_ABI_VERSION` documentation
4. remove subdirectory "fltk" (Makefile section)

> If this seems OK, I'll move forward and modify the CMP section on macro variables to simply refer to this.

OK, seems good to me.

**Comment by Albrecht-S** [comment]
Some thoughts: Eventually I'd like to integrate all or most of our developer documentation in the separate "Developer Docs" document. This includes all README files concerning developers. Most of the other README files should be merged into the "User Docs" document (FLTK doxygen docs). There are two reasons for this:

1. Less info in PHP files on the web site which are difficult to maintain (only by a few maintainers, you know who these are) ;-)
2. We need to consolidate our many README files in better structured documents (the user docs and the developer docs)

There may be a small number of (or only one) remaining README file(s) for users to start with. This README file might end up as README.md (with markdown) which will be shown on the main GitHub page as usual for OS projects.

Everything that makes this transition easier is a good step.

**Comment by Albrecht-S** [comment]
I just noticed that `WIN32` is still used in the TOC, please change to `_WIN32` (with underbar).

**Comment by erco77** [comment]
OK: applied your patch and made the WIN32 -> _WIN32 TOC change.
I added the file to my Fl_Terminal branch, so it'll sneak in as part of that whenever I get the OK to do a PR for that.

[I may have misread your sugg to move stuff from README -> dox.. but just in case, we should be careful what we move. I agree some developer docs that don't involve a lot of command line related stuff, like design documents, should be moved out to doxygen docs. Anyway, just outlining what I want to make sure stays README txt files in quotes here:]
> Regarding moving the e.g. cmake README file into doxygen, I dunno; I think users trying to build FLTK from the command line would prefer to refer to a README, rather than try to find what they're looking for in what is fast becoming a 1000 page document..! Also, what if they don't have doxygen and are working with a new release; they wouldn't be able to read the instructions.
> 
> I myself constantly am grepping what I need from the README.CMake.txt to find the macro variables I'm looking for so I can copy/paste them into the cli, e.g. `grep Debug` to get that debug build macro name, and `grep EXAMPLES` to get the macro for building the examples.

Certainly some stuff could and should be moved, or combined and reduced, and as we discussed, a single top level README.txt that refers to them all, small one line descriptions of each, like an index to the README's.

**Comment by Albrecht-S** [comment]
I agree that a certain amount of README files is good and maybe needed for a beginning. Maybe one? SCNR, I'm kidding.

But I also see that our current README's contain duplicated information and to build on Linux with CMake you need to consult two different README's. If you want to use Wayland you need to use a third README file. This would be much easier and more consistent (w/o duplicated and eventually outdated information) if this would be written in the doxygen user docs with links to the appropriate chapters. Currently we have 11 README files, and this new one would be no. 12.

Regarding starting a new project and needing README files to copy and paste:
1. we have online docs
2. if you don't like online docs you can download docs as HTML or PDF and use them locally

I admit that the size (number of pages) of the online docs might be an issue, but I believe that this is a minor problem and can be solved by good structure. If you need to setup a new project you should find the instructions in a specific chapter in the HTML or PDF docs [1] and copy/paste works with these docs as well. You could even jump from `Unix/Linux` to `CMake` to `Wayland` setup docs quickly by links.

[1] This chapter is not yet written, we still have only the old and outdated configure/make docs!

**Comment by erco77** [comment]
The CMP might be a good thing to move into doxygen, e.g. documentation/source/cmp.dox

If we want to do that, it'd be a good idea for me to do that now, instead of merging into the PHP source.
**Should I do that? (Y/N)**

Should be straight forward to convert to doxygen with headings + subheadings. Perhaps generated as a separate document, not part of the FLTK source docs. This should be able to be a simple target to include in our default Makefile.

**Comment by Albrecht-S** [comment]
> The CMP might be a good thing to move into doxygen, ...
> Should I do that? (Y/N)

YES, please! That's an excellent idea.

I just took a look over the full CMP and it seems that almost all of this could be the starting chapter(s) of the new "FLTK Developer Documentation" (introduction + coding standard).

There are parts that might better be restructured (separated, moved to other chapters) like dealing with bug reports and the release process, but everything would be good to move to the doxygen developer docs. Then other devs could edit this as usual, and we could remove the "cmp.php" file completely (or leave it as a short info with a link to the new docs).

As a start I agree with converting the entire document to one doxygen file as you proposed, e.g. "documentation/src/cmp.dox". Once we have it in git we can edit and restructure it as we need.

If you're going to do this (will you?) then I'd wish that you convert some of the HTML stuff that doxygen **can** understand to "doxygen syntax" (markup) that's easier to understand and edit, particularly enumerations (`<ul><li>...</li></ul>`). This should be done before the first checkin. Such lists would be a single hyphen (dash) per list item (numbered lists would be `'-#'`. This would simplify the new docs.

**Comment by erco77** [comment]
OK, good.

Yes, I'll see about splitting the CMP related stuff (releases vs. github issue management, etc) into separate dox files, such as maybe:

- cmp-coding-style.dox -- just coding style and doxygen tips
- developer-github.dox -- issue management/commit/etc
- developer-release.dox  -- release process
- developer-release-numbers.dox (can be #included by both public and developer docs)

Perhaps the doxygen docs can either (a) #include, or (b) simply refer by name to the README.xyz.txt files (for Makefile targets, CMake instructions, etc)

And yes, replacing html code as much as possible with doxygen markup. (I think some things like underlining and italics might still have to be done in html)

**Comment by Albrecht-S** [comment]
+1 :+1: 

> I think some things like underlining and italics might still have to be done in html

... at least if you need multi-word sequences, for instance bold `\b` can mark only one word, otherwise you need HTML. I'm not sure about italics and underline. There's AFAICT no alternative for `<kbd>..</kbd>` as well.

> Perhaps the doxygen docs can either (a) #include, or (b) simply refer by name to the README.xyz.txt files (for Makefile targets, CMake instructions, etc)

That depends...
- Makefile targets **for the library** should be directly in the developer dox
- `CMake` instructions (for building programs with FLTK) are in user (not dev.) dox
- we don't have `CMake` instructions for building the library yet [1]

[1] I don't think that we can cover everything we need to build the library (the same is true for `configure.ac`), but we could maybe add `CMake` formatting instructions (indent etc.) and "how to find a particular library or function" in `CMake`. Such information would go into the dev dox.

**Comment by erco77** [comment]
> Makefile targets for the library should be directly in the developer dox

Mmm, but users need to know them too. Users who downloaded the lib and want to build the toolkit might be headless, and would expect e.g. README.txt, BUILD.txt or INSTALL.txt to guide them. I myself as a user did not know about targets like distclean until I became a dev.. feel like I should have known that one sooner!

The people building FLTK tend to be either app devs trying out fltk for the first time, or systems admins installing it because it's needed to build other tools, or at the request of app devs. At the beginning they won't be reading developer docs (if ever), so best not to hide that stuff from non-devs.

So my thinking is, anything that /should/ be in the README txt files should be referenced by the doxygen docs (or pulled in via an #include directly from the README's at doxygen build time, to prevent duplicate sources of the info)

**Comment by erco77** [comment]
> There's AFAICT no alternative for \<kbd>..\</kbd> as well.

If \<kbd> is like \<tt> (I use the latter in html), I think backticks is the doxygen equivalent, much like markup here in githib.

**Comment by Albrecht-S** [comment]
OK, I misunderstood what you meant with "Makefile targets". I thought you meant how to **create** Makefile targets of the library, e.g. a specific lib...

I fully agree that the existing Makefile targets as well as the CMake options are relevant for users and should be documented in the user docs and/or README files (not in developer docs for sure).

My understanding of developer docs is that these are exclusively for developers **of** the FLTK library and for other interested people, potential contributors etc..

**Comment by erco77** [comment]
Doxygen-ifying the coding standards has been done and is in git fltk master.

Closing this issue here on my fork - all continuing conversation about this issue can continue on fltk as issue #1346, and that issue refers to this (now closed) issue in case there's needed info here.
